frontend/i_types.py

```diff
@@ -0,0 +1,342 @@
+    """
+
+    def __init__(self, name, context=Context()):
+        self.name = name
+        self.context = context
+
+    def is_subtype(self, t):
+        if not isinstance(t, TClass):
+            return False
+        return subtypes_tree.reachable(self.name, t.name)
+
+class TNone(Type):
+
+    def is_subtype(self, t):
+        return True
+
+    def get_name(self):
+        return "None"
+
+# TODO: Numerics should be subtype of class
+class TBool(TClass):
+    def __init__(self):
+        self.name = "bool"
+
+    def get_name(self):
+        return self.name
+
+
+class TInt(TClass):
+    def __init__(self):
+        self.name = "int"
+
+    def get_name(self):
+        return self.name
+
+
+class TFloat(TClass):
+
+    def __init__(self):
+        self.name = "float"
+
+    def get_name(self):
+        return self.name
+
+
+class TString(TClass):
+
+    def __init__(self):
+        self.name = "str"
+
+    def get_name(self):
+        return self.name
+
+
+class TBytesString(TClass):
+
+    def __init__(self):
+        self.name = "bytes"
+
+    def get_name(self):
+        return self.name
+
+
+class TList(TClass):
+    """Type given to homogeneous lists.
+
+    Attributes:
+        type (Type): Type of the list elements
+    """
+
+    def __init__(self, t):
+        self.type = t
+        self.name = "list"
+
+    def is_subtype(self, t):
+        if isinstance(t, TClass) and t.name == "object":
+            return True
+        return isinstance(t, TList) and type(self.type) is type(t.type)
+
+    def get_name(self):
+        return "{}[{}]".format(self.name, self.type.get_name())
+
+class TTuple(TClass):
+    """Type given to a tuple.
+
+    Attributes:
+        types ([Type]): Types of the tuple elements.
+    """
+
+    def __init__(self, t):
+        self.types = t
+        self.name = "tuple"
+
+    def is_subtype(self, t):
+        if isinstance(t, TClass) and t.name == "object":
+            return True
+        if not isinstance(t, TTuple):
+            return False
+        if len(self.types) != len(t.types):
+            return False
+        for i in range(len(self.types)):
+            if not self.types[i].is_subtype(t.types[i]):
+                return False
+        return True
+
+    def get_name(self):
+        types_names = [t.get_name() for t in self.types]
+        return "{}({})".format(self.name, ",".join(types_names))
+
+    def get_possible_tuple_slicings(self):
+        """Returns a union type of all possible slicings of this tuple
+
+		For example:
+			t = (1, "string", 2.5)
+
+			t.get_possible_tuple_slicings() will return the following
+			Union{Tuple(Int), Tuple(Int,String), Tuple(Int,String,Float), Tuple(String), Tuple(String,Float), Tuple(Float), Tuple()}
+
+		"""
+        slicings = {TTuple([])}
+        for i in range(len(self.types)):
+            for j in range(i + 1, len(self.types) + 1):
+                slicings.add(TTuple(self.types[i:j]))
+        return UnionTypes(slicings)
+
```

Comment:
Again, I would suggest to follow the documentation (https://docs.python.org/3.5/reference/datamodel.html) more closely. For instance, looking at how 'sequences' are organized: 'string', and 'tuple' are types of 'immutable sequences', and 'list' is a type of 'mutable sequence'. Since all sequence types support common functionalities (len(), slicing, ...) it might make sense to have a common superclass 'sequence' and two subclasses 'immutable sequence' and 'mutable sequence' and share the code accordingly.